pkg/build/api/types.go

```diff
@@ -20,6 +20,30 @@ const (
 	DefaultDockerLabelNamespace = "io.openshift."
 )
 
+const (
+	// SecretTypeSource contains data needed to authenticate against SCM servers.
+	//
+	// SecretTypeSource keys are devided into these option groups:
+	//   Basic authentication group. Requires at least one of fields:
+	//     - Secret.Data["username"] - username used to authenticate against SCM server
+	//     - Secret.Data["password"] - password or token needed to authenticate against SCM server
+	//     - Secret.Data["ca.crt"] - CA certificate needed to authenticate against SCM server
+	//   SSH authentication group. Required field:
+	//     - Secret.Data["ssh-privatekey"] - private SSH key needed to authenticate against SCM server
+	SecretTypeSource kapi.SecretType = "Source"
```

Comment:
Username/password are generally useful, ssh private key is generally
useful.  If gitconfig is a third type (namespaced to Openshift) and can
also be added to both ssh private key and username password, then the
builder simply needs to take those into account.  I don't believe the
builder needs to support more than one username, private key, or token
secret for source.

We should validate - but we've got the wrong pattern for validation in Kube
to support third party extension (unless that validation can be expressed
as a schema).

On Sep 9, 2015, at 8:53 AM, David Eads <notifications@github.com> wrote:

In pkg/build/api/types.go
<https://github.com/openshift/origin/pull/4433#discussion_r39037676>:

> @@ -20,6 +20,30 @@ const (
>  	DefaultDockerLabelNamespace = "io.openshift."
>  )
>
> +const (
> +	// SecretTypeSource contains data needed to authenticate against SCM servers.
> +	//
> +	// SecretTypeSource keys are devided into these option groups:
> +	//   Basic authentication group. Requires at least one of fields:
> +	//     - Secret.Data["username"] - username used to authenticate against SCM server
> +	//     - Secret.Data["password"] - password or token needed to authenticate against SCM server
> +	//     - Secret.Data["ca.crt"] - CA certificate needed to authenticate against SCM server
> +	//   SSH authentication group. Required field:
> +	//     - Secret.Data["ssh-privatekey"] - private SSH key needed to authenticate against SCM server
> +	SecretTypeSource kapi.SecretType = "Source"

You can't because we have to run against external Kube. Kube should have
more secret types and not reject types it doesn't know about. Kube secret
types should be more flexible and allow unrelated secret data in them.

Kube doesn't reject unknown secret types for exactly this reason. Having a
way to add additional validation for custom secrets types is strictly
better than not having that validation, even if its not always available.

Moving this secret type upstream seemed unlikely given its narrow usage and
secrets cannot have multiple types.

—
Reply to this email directly or view it on GitHub
<https://github.com/openshift/origin/pull/4433/files#r39037676>.